text/0000-replace-blacklist-whitelist.md

```diff
@@ -0,0 +1,70 @@
+- Start Date: 2020-06-15
+- Relevant Team(s): Ember CLI, Learning
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Replace terms blacklist & whitelist in Ember CLI
+
+## Summary
+
+Ember.js prides itself (rightly) on being an inclusive framework. To further improve on this, we should remove the terms "blacklist" and "whitelist" with more neutral replacements.
+
+## Motivation
+
+The terms "blacklist" and "whitelist" are considered racially insenstive. These terms are used in Ember CLI for rather advanced functionality, which should make providing a better alternative naming and deprecating the current one rather easy.
+
+## Detailed design
+
+You can use `blacklist` and `whitelist` to include/exclude certain addons from your build:
+
+https://ember-cli.com/user-guide/#whitelisting-and-blacklisting-assets
+
+We should replace these with `exclude` or `include`:
+
+```js
+// Before
+let app = new EmberApp(defaults, {
+  addons: {
+    blacklist: ["fastboot-app-server"],
+  },
+});
+
+// After
+let app = new EmberApp(defaults, {
+  addons: {
+    exclude: ["fastboot-app-server"],
+  },
+});
+
+// Before
+let app = new EmberApp(defaults, {
+  addons: {
+    whitelist: ["ember-cli-sass"],
+  },
+});
+
+// After
+let app = new EmberApp(defaults, {
+  addons: {
+    include: ["ember-cli-sass"],
+  },
+});
+```
+
+The old keys (`blacklist` and `whitelist`) should be aliased to the new ones, and show a deprecation warning. Functionally, nothing changes.
```

Comment:
There is a lot of material that is available about the subliminal messaging that is pervasive in American culture whose underlying message is "white is good" and that has made its way into a lot of things as "the norm".

While the original developer who first wrote "whitelist" and "blacklist" was perhaps not a racist (but rather using the words that had been made social norms by a white majority), we have enough public awareness to choose an elevated, more inclusive way of doing things for Ember projects.

Reply:
I am not sure, as far as I see this only affects apps, not addons. And it will only ever appear exactly once at one place, and can be immediately changed without much thought.

I am also OK with not deprecating this now, but then, when would we? Will it be less churn if it it deprecated later? :shrug: But I also have no strong feelings otherwise. What timeline would you think about for deprecation?

Reply:
As long as this only affects apps (and addons not getting upgraded won't produce deprecations in consuming apps) I feel the noise is minimal and the cost of deprecating is the same as or less than the cost of maintaining both keys.

Reply:
@MelSumner 

> There is a lot of material that is available about the subliminal messaging that is pervasive in American culture whose underlying message is "white is good" and that has made its way into a lot of things as "the norm".
> 
> While the original developer who first wrote "whitelist" and "blacklist" was perhaps not a racist (but rather using the words that had been made social norms by a white majority), we have enough public awareness to choose an elevated, more inclusive way of doing things for Ember projects.

How does this have to do with what I said at all? We all really need to make sure to stay focused on the actual proposal here...

@mydea

> I am not sure, as far as I see this only affects apps, not addons. And it will only ever appear exactly once at one place, and can be immediately changed without much thought.

As I mentioned above, we should not introduce churn on existing application developers for this. If they want to migrate off of the previous terms: seems completely fine and we will provide a nice alternative (this RFC!). 

@Gaurav0 

> I feel the noise is minimal and the cost of deprecating is the same as or less than the cost of maintaining both keys.

The cost of deprecation is definitely more overall. It requires:

* adding support for both fields
* adding the deprecation notice itself
* requires all applications that use these APIs to modify their application code

Whereas supporting both fields has only the first. Now, in general, ember-cli developers (and addon authors) are commonly willing to absorb churn like the one proposed in this RFC in order to make the lives of our users better (hence the willingness to support a version of this RFC), but instead of reducing churn for our consumers by doing the right thing internally adding a deprecation _intentionally_ adds **exactly** the churn we intend to avoid.

Reply:
Ember has an existing policy for breaking changes, following that process seems like a tried tested and true approach.

Reply:
I know y'all have good intentions, so just want to say... IMO it's not a requirement to make this transition easy or comfortable or following the appropriate schedule (whatever that is), if it's happening for the right reasons. I honestly can't see a good reason _not_ to immediately deprecate.

Reply:
If this RFC is accepted I think it is out of question that it will _eventually_ be deprecated (as we would not want to indefinitely keep two option keys for the same thing). So as far as I see, the actual question is _when_ should it be deprecated.

As this is not a gradual transition, but a one-time change always done in the same place, at most once per application, I do not see the benefit in postponing this. Thinking about churn, there are these possible scenarios:

**Case A: We ship the change but deprecate it later**
**Option 1**: I read about the change in the release docs, so I go into my app to check if I've been using this, and if so, change it.
**Option 2**: I either don't realize this has been changed, or I do not want to change it right now. So I do nothing initially. X months later it is deprecated, so I do Option 1 then.

**Case B: We ship the change and deprecate it immediately**
I get notified if I've been using this and change it in my app.

I don't see anybody using a deprecation-free transition time of X months to slowly change this, as this cannot be slowly changed - either you update it in your app, or you don't.

Overall, I am not necessarily against deprecating it later, I simply don't see how it would help anybody. But I'm also happy to include a different deprecation strategy if you would prefer. What time horizon would make you feel more comfortable, deprecation wise?

Reply:
As a general principle, deprecation RFCs don't try to guess the exact timing of when the deprecations will be released. Instead, they describe a phasing strategy for the deprecations, sometimes along with criteria that should be met for each stage.

In other words, deprecation RFCs are about the details of issuing deprecations, not about the exact versions when the deprecations will appear in the codebase.

Reply:
> In other words, deprecation RFCs are about the details of issuing deprecations, not about the exact versions when the deprecations will appear in the codebase.

Yes, exactly. My point was that I am opposed to immediate deprecation. It should be a phased transition plan (introduce new API, deprecate in a future version, make an error upon deprecation removal), like we normally do.

Reply:
OK, I am fine with that - so I will remove the mention of the deprecation.

To cap this off, my thoughts on this: IMHO this change is a bit different from most other changes. Usually we would introduce a new API with changed behavior, and give devs time to migrate to the new behavior before deprecating it. However, in this case it is a one-to-one replacement of a rather niche functionality that is used at most once per application. That was my reasoning for proposing to deprecate it immediately, instead of requiring another RFC sometime later to deprecate it. IMHO it will potentially even introduce _more_ churn as new users might still be using the old terms (and not even know there are new ones), e.g. by copy-pasting code from somewhere or similar. But I respect your desire to follow the usual process :+1:

Reply:
seems like its deprecated as soon as the new fields are available. whether you warn people about the deprecation is another thing.

i'm using deprecation in the practical sense of "this field will go away in favor of this other thing".

$0.02